content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,114 @@
+
+
+Clang-doc previously had test’s which did not test the full scope of the tool especially with regards to the HTML or Markdown output of the tool. In order to make sure the optimization experiments were not accidentally breaking the tool, it was necessary to add more end to end tests. 
+In summary I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 
+
+
+### Improve clang-doc’s performance by 1.58 times 
+
+Internally the way clang-doc works is by leveraging libtooling's ASTVisitor class to parse the declaration in each project and serializing it into an internal format, which gets deserialized later when we output the final format. The bottleneck in clang-doc lied in the way clang-doc was doing redundant work when it was visiting each declaration. 
+
+For example if we had a the following project: 
+
+```cpp
+//File: Base.cpp
+
+class Base {}
+```
+
+
+```cpp
+//File: A.cpp
+
+class A : public Base {}
+```
+
+```cpp
+//File: B.cpp
+
+class B : public Base {}
+```
+
+In this case the ASTVisitor invoke by clang-doc would visit serialized the Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 
+
+The optimization ended up being a simple memoization dictionary which kept track of a list of declaration that clang-doc had visited. 
+
+Here is a plot of the benchmarking numbers:
+
+<div style="margin:0 auto;">
+  <img src="/img/clang-doc-benchmark-numbers.png"><br/>
+</div>
+
+
+### Added Template Mustache HTML Backend
+
+Clang-doc originally used an ad-hoc method of generating HTML, in this project I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however upon further discussion it was decided that the complexity of designing and implementing a new templating language was too much. 
+A LLVM community member suggested I implement a mustache as templating language. 
+Mustache was the ideal templating language, since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 
+In terms of engineering wins, this library was able to cut the direct down on HTML backend significantly dropping 500 lines of code. This library was also designed for general purpose use around LLVM, since there are numerous places in LLVM where various tools generate html in its own way. Using the mustache templating library would be a nice way to standardize the codebase. 
```

Comment:
What does "cut the direct down on HTML backed" mean? do you mean that we were able to drop 500 lines of C++ from clang-doc? or something else.